

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>2.10. Loading Blast results &mdash; machado 0.1.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="2.11. Loading InterproScan results" href="load_interproscan.html" />
    <link rel="prev" title="2.9. Loading Feature Publication File" href="load_feature_publication.html" /> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> machado
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">1. Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="dataload.html">2. Data loading</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="load_ontologies.html">2.1. Ontologies</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_taxonomy.html">2.2. Taxonomy</a></li>
<li class="toctree-l2"><a class="reference internal" href="insert_organism.html">2.3. Organism</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_publication.html">2.4. Publication</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_fasta.html">2.5. FASTA</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_gff.html">2.6. GFF</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_feature_annotation.html">2.7. Feature annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_feature_sequence.html">2.8. Feature sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_feature_publication.html">2.9. Feature publication</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">2.10. Blast</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_interproscan.html">2.11. InterproScan</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_orthomcl.html">2.12. OrthoMCL</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_rnaseq.html">2.13. RNA-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_coexpression.html">2.14. Coexpression</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="visualization.html">3. Visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="diagrams.html">4. Diagrams</a></li>
<li class="toctree-l1"><a class="reference internal" href="models.html">5. Models</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">machado</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="dataload.html">2. Data loading</a> &raquo;</li>
        
      <li>2.10. Loading Blast results</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/load_blast.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="loading-blast-results">
<h1>2.10. Loading Blast results<a class="headerlink" href="#loading-blast-results" title="Permalink to this headline">¶</a></h1>
<p>In order to load a BLAST xml result file, both the query and the subject records must be previously stored (see below). The current version was tested for loading BLAST analysis on <strong>proteins</strong>.</p>
<div class="section" id="load-blast-subject-records">
<h2>2.10.1. Load BLAST subject records<a class="headerlink" href="#load-blast-subject-records" title="Permalink to this headline">¶</a></h2>
<p>In case you did a BLAST against a multispecies protein database, like NCBI’s nr or Uniprot’s trembl or swissprot, you need to load previously all subject matches before loading the result itself. To do so, use the following command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python manage.py load_similarity_matches --file blast_result.xml --format blast-xml
</pre></div>
</div>
<ul class="simple">
<li>Loading this file can be faster if you increase the number of threads (–cpu).</li>
</ul>
</div>
<div class="section" id="load-blast">
<h2>2.10.2. Load BLAST<a class="headerlink" href="#load-blast" title="Permalink to this headline">¶</a></h2>
<p>If all queries and subjects are already loaded you can run the following command to load a BLAST xml result:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python manage.py load_similarity --file blast_result.xml --format blast-xml --so_query polypeptide --so_subject protein_match --program diamond --programversion <span class="m">0</span>.9.24 --organism_query <span class="s1">&#39;Oryza sativa&#39;</span> --organism_subject <span class="s1">&#39;multispecies multispecies&#39;</span>
</pre></div>
</div>
<ul class="simple">
<li>Loading this file can be faster if you increase the number of threads (–cpu).</li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python manage.py load_similarity --help
</pre></div>
</div>
<table border="1" class="docutils">
<colgroup>
<col width="14%" />
<col width="86%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td>–file</td>
<td>Blast XML file *</td>
</tr>
<tr class="row-even"><td>–format</td>
<td>blast-xml *</td>
</tr>
<tr class="row-odd"><td>–so_query</td>
<td>Query Sequence Ontology term. (eg. assembly, mRNA, polypeptide) *</td>
</tr>
<tr class="row-even"><td>–so_subject</td>
<td>Subject Sequence Ontology term. (eg. protein_match if loading a BLAST result) *</td>
</tr>
<tr class="row-odd"><td>–organism_query</td>
<td>Query’s organism name. eg. ‘Oryza sativa’. Cannot be ‘multispecies’ *</td>
</tr>
<tr class="row-even"><td>–organism_subject</td>
<td>Subject’s organism name eg. ‘Oryza sativa’. Put ‘multispecies multispecies’ if using a multispecies database. *</td>
</tr>
<tr class="row-odd"><td>–program</td>
<td>Program *</td>
</tr>
<tr class="row-even"><td>–programversion</td>
<td>Program version *</td>
</tr>
<tr class="row-odd"><td>–name</td>
<td>Name</td>
</tr>
<tr class="row-even"><td>–description</td>
<td>Description</td>
</tr>
<tr class="row-odd"><td>–algorithm</td>
<td>Algorithm</td>
</tr>
<tr class="row-even"><td>–cpu</td>
<td>Number of threads</td>
</tr>
</tbody>
</table>
<p>* required fields</p>
</div>
<div class="section" id="remove-file">
<h2>2.10.3. Remove file<a class="headerlink" href="#remove-file" title="Permalink to this headline">¶</a></h2>
<p>If, by any reason, you need to remove a Blast result set you should use the command <em>remove_analysis</em>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python manage.py remove_analysis --help
</pre></div>
</div>
<ul class="simple">
<li>This command requires the analysis name (Analysis.sourcename)</li>
</ul>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="load_interproscan.html" class="btn btn-neutral float-right" title="2.11. Loading InterproScan results" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="load_feature_publication.html" class="btn btn-neutral" title="2.9. Loading Feature Publication File" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Embrapa.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0.1.0',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="_static/language_data.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>